File: Java/Beginner/E_ShortestPath/boj_1916.java
package Beginner.E_ShortestPath;

import java.io.BufferedReader;
import java.io.IOException;
import java.io.InputStreamReader;
import java.util.*;

public class boj_1916 {

    static StringTokenizer st;
    static List<Node>[] grp;
    static final int INF = Integer.MAX_VALUE;

    static class Node implements Comparable<Node> {
        int end;
        int weight;

        public Node(int end, int weight) {
            this.end = end;
            this.weight = weight;
        }

        @Override
        public int compareTo(Node o) {
            return this.weight - o.weight;
        }
    }

    private static int dijkstra(int start, int end) {
        PriorityQueue<Node> pq = new PriorityQueue<>();
        int[] distance = new int[grp.length];
        Arrays.fill(distance, INF);

        pq.add(new Node(start, 0));
        distance[start] = 0;

        while (!pq.isEmpty()) {
            Node tmp = pq.poll();
            int now = tmp.end;
            int dist = tmp.weight;

            if (distance[now] < dist)
                continue;

            for (Node node : grp[now]) {
                int end_vertex = node.end;
                int weight = node.weight;
                int cost = distance[now] + weight;

                if (distance[end_vertex] > cost) {
                    distance[end_vertex] = cost;
                    pq.add(new Node(end_vertex, cost));
                }
            }
        }

        return distance[end];
    }

    public static void main(String[] args) throws IOException {
        BufferedReader br =
                new BufferedReader(new InputStreamReader(System.in));

        int n = Integer.parseInt(br.readLine());
        int m = Integer.parseInt(br.readLine());

        grp = new ArrayList[n + 1];
        Arrays.setAll(grp, elem -> new ArrayList<>());
        /*
        for (int i = 0; i <= n; i++)
            grp[i] = new ArrayList<>();
        */

        while (m-- > 0) {
            st = new StringTokenizer(br.readLine());
            int u = Integer.parseInt(st.nextToken());
            int v = Integer.parseInt(st.nextToken());
            int w = Integer.parseInt(st.nextToken());

            grp[u].add(new Node(v, w));
        }

        st = new StringTokenizer(br.readLine());
        int start = Integer.parseInt(st.nextToken());
        int end = Integer.parseInt(st.nextToken());

        System.out.print(dijkstra(start, end));

        br.close();
    }
}

File: Kotlin/src/main/kotlin/beginner/e_ShortestPath/boj_1916.kt
package beginner.e_ShortestPath

import java.util.*


private data class node(val end: Int, val weight: Int): Comparable<node> {
    override fun compareTo(other: node): Int {
        return this.weight - other.weight
    }
}

private fun solve(grp: MutableList<PriorityQueue<node>>, start: Int, end: Int): Int {
    val pq = PriorityQueue<node>()
    val distance = IntArray(grp.size){Int.MAX_VALUE}

    pq.add(node(start, 0))
    distance[start] = 0

    while (pq.isNotEmpty()) {
        val (now, dist) = pq.poll()

        if (distance[now] < dist)
            continue

        for ((next, weight) in grp[now]) {
            val cost = distance[now] + weight

            if (distance[next] > cost) {
                distance[next] = cost
                pq.add(node(next, cost))
            }
        }
    }

    return distance[end]
}

fun boj_1916() = with(System.`in`.bufferedReader()) {
    val n = readLine().toInt()
    val m = readLine().toInt()

    val grp = MutableList(n){ PriorityQueue<node>() }

    repeat(m) {
        val (u, v, w) = readLine()
            .split(" ")
            .map { it.toInt() }

        grp[u - 1].add(node(v - 1, w))
    }

    val (start, end) = readLine()
        .split(" ")
        .map { it.toInt() }

    print(solve(grp, start - 1, end - 1))
}